File: 500. Keyboard Row.cpp
class Solution {
public:
    // hash table: O(mn)time, O(1)space, 1.85%
    vector<string> findWords(vector<string>& words) {
        vector<string> ret;
        if(words.empty()) return ret;
        unordered_map<int, string> mapping
            = {{1, "qwertyuiop"}, {2, "asdfghjkl"}, {3, "zxcvbnm"}};

        for(int i = 0; i < words.size(); i++){
            int tag;
            if(!words[i].empty()){
                for(auto m: mapping){
                    if(m.second.find(tolower(words[i][0])) != string::npos){
                        tag = m.first;
                        break;
                    }
                }
            }
            bool isValid = true;
            for(int j = 1; j < words[i].size(); j++){
                string s = mapping[tag];
                if(s.find(tolower(words[i][j])) == string::npos){
                    isValid = false;
                    break;
                }
            }
            if(isValid) ret.push_back(words[i]);
        }
        return ret;
    }

    // regular expression
    vector<string> findWords(vector<string>& words) {

    }

};

File: 500. Keyboard Row.py
class Solution(object):
    
    # Brute force: O(mn) time, O(m) space, 68.37%
    def findWords(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        letters = ['qwertyuiop', 'asdfghjkl', 'zxcvbnm']
        rows, ret = dict(), list()
        for index, letter in enumerate(letters):
            rows.update({c: index for c in letter})

        for word in words:
            wordLower = word.lower()
            rowNum, idx = [rows[c] for c in wordLower], rows[wordLower[0]]
            same = True
            for c in wordLower:
                if not rows[c] == idx:
                    same = False
                    break
            if same: ret.append(word)
        return ret
        
    # Set: O(mn) time, O(m) space, 13.44%
    def findWords(self, words):
        lines, ret = [set('qwertyuiop'), set('asdfghjkl'), set('zxcvbnm')], []
        for word in words:
            w = set(word.lower())
            for line in lines:
                if w.issubset(line):
                    ret.append(word)
        return ret

    # Set: O(mn) time, O(m) space, 52.56%
    def findWords(self, words):
        line1, line2, line3, ret = set('qwertyuiop'), set('asdfghjkl'), set('zxcvbnm'), []
        for word in words:
            w = set(word.lower())
            if line1 & w == w or line2 & w == w or line3 & w == w:
                ret.append(word)
        return ret
